File: cloud/filestore/public/sdk/python/client/credentials.py
import grpc
import os


_MAX_CERT_FILE_SIZE = 8 * 1024 * 1024


class ClientCredentials(object):

    def __init__(
            self,
            root_certs_file=None,
            cert_file=None,
            cert_private_key_file=None,
            auth_token=None):

        self.root_certs_file = root_certs_file
        self.cert_file = cert_file
        self.cert_private_key_file = cert_private_key_file
        self.auth_token = auth_token

    def get_ssl_channel_credentials(self):
        credentials = grpc.ssl_channel_credentials(
            root_certificates=self._read_file_optional(self.root_certs_file),
            private_key=self._read_file_optional(self.cert_private_key_file),
            certificate_chain=self._read_file_optional(self.cert_file),
        )

        if self.auth_token is not None:
            credentials = grpc.composite_channel_credentials(
                credentials,
                grpc.access_token_call_credentials(self.auth_token),
            )

        return credentials

    @staticmethod
    def _read_file_optional(filename=None):
        if filename is None:
            return None

        st = os.stat(filename)
        if st.st_size > _MAX_CERT_FILE_SIZE:
            raise Exception("file too large")

        with open(filename, "rb") as f:
            return f.read()

File: cloud/filestore/public/sdk/go/client/credentials.go
package client

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"errors"
	"fmt"
	"os"

	"golang.org/x/oauth2"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/oauth"

	storage_grpc "github.com/ydb-platform/nbs/cloud/storage/core/go/grpc"
)

////////////////////////////////////////////////////////////////////////////////

const authHeader = "authorization"
const authMethod = "Bearer"

type ClientCredentials struct {
	RootCertsFile      string
	CertFile           string
	CertPrivateKeyFile string
	// TlsProvider, when set, takes precedence over RootCertsFile,
	// CertFile, and CertPrivateKeyFile.
	TlsProvider TlsConfigProvider
	AuthToken   string
	IAMClient   TokenProvider
}

type TlsConfigProvider = storage_grpc.TlsConfigProvider

type TokenProvider interface {
	Token(ctx context.Context) (string, error)
}

type grpcTokenProvider struct {
	provider TokenProvider
}

func (p *grpcTokenProvider) GetRequestMetadata(ctx context.Context, _ ...string) (map[string]string, error) {
	token, err := p.provider.Token(ctx)
	if err != nil {
		return nil, err
	}
	return map[string]string{authHeader: authMethod + " " + token}, nil
}

func (p *grpcTokenProvider) RequireTransportSecurity() bool {
	return false
}

func (creds *ClientCredentials) GetSslChannelCredentials() ([]grpc.DialOption, error) {
	var transportCredentials credentials.TransportCredentials
	// A typed-nil TlsProvider is a configuration error. Let the constructor
	// reject it instead of silently falling back to the certificate files.
	if creds.TlsProvider != nil {
		var err error
		transportCredentials, err = storage_grpc.NewGrpcClientTransportCredentials(
			creds.TlsProvider,
		)
		if err != nil {
			return nil, err
		}
	} else {
		cfg := &tls.Config{}
		if creds.CertFile != "" {
			cert, err := tls.LoadX509KeyPair(creds.CertFile, creds.CertPrivateKeyFile)
			if err != nil {
				return nil, fmt.Errorf(
					"failed to load client certificate/key: %s",
					err.Error(),
				)
			}

			cfg.Certificates = []tls.Certificate{cert}
		}

		if creds.RootCertsFile != "" {
			pem, err := os.ReadFile(creds.RootCertsFile)
			if err != nil {
				return nil, fmt.Errorf("failed to read root cert file: %s", err.Error())
			}

			pool := x509.NewCertPool()
			ok := pool.AppendCertsFromPEM(pem)
			if !ok {
				return nil, errors.New("failed to parse PEM")
			}

			cfg.RootCAs = pool
		}

		transportCredentials = credentials.NewTLS(cfg)
	}

	opts := []grpc.DialOption{
		grpc.WithTransportCredentials(transportCredentials),
	}

	if creds.AuthToken != "" {
		token := oauth2.Token{AccessToken: creds.AuthToken}
		opts = append(opts, grpc.WithPerRPCCredentials(oauth.NewOauthAccess(&token)))
	}

	if creds.IAMClient != nil {
		opts = append(opts, grpc.WithPerRPCCredentials(&grpcTokenProvider{provider: creds.IAMClient}))
	}

	return opts, nil
}
